packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m

```diff
@@ -237,7 +237,7 @@ + (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
 
 + (SKPaymentDiscount *)getSKPaymentDiscountFromMap:(NSDictionary *)map
                                          withError:(NSString **)error {
-  if (!map || map.count <= 0) {
+  if (!map || [map isKindOfClass:[NSNull class]] || map.count <= 0) {
```

Comment:
> when I was buying a product without discount, it was `NULL`

You mean it was `NSNull`, presumably, given the code you added. `NULL` is a typedef for `0`, and is essentially the C equivalent of `nil`. `NSNull` is not `nil`/`NULL`, it is singleton class.

> but it did not stop at the `!map`

That is expected; a pointer that is pointing to an `NSNull` instance is not `nil`.

> it went to the next check `map.count <= 0` and crashed.

Presumably the crash was an `unrecognized selector sent to instance` exception, because `count` is not a method of `NSNull`.

> So here I do not try to check for a type violation

Yes, that is exactly what you are checking for. You are taking a pointer of type `NSDictionary` and checking whether it contains an `NSNull` instance instead of an `NSDictionary` instance. Which would be a type violation.

> As for if it should be done here or in the caller. Why it is better in the caller?

This method takes an `NSDictionary*`, not an `id`. It is better in the caller because it is a type violation to pass an argument to this method that is not an `NSDictionary`, and an instance of `NSNull` **is not** an `NSDictionary`.

The comparable Dart code is:

```
class NSNull {
}

void methodHandler(dynamic args) {
  expectsAMap(args);
}

void expectsAMap(Map? map) {
  if (map != null) {
    print(map.length);
  }
}

void main() {
  NSNull nullArg = NSNull();
  methodHandler(nullArg);
}
```

The bug is this Dart code is `methodHandler` passing `args` to `expectsAMap` without checking whether it's a `Map`, not when `expectsAMap` uses `map` as if it were its declared type.

> Shouldn't the method take care of the argument's validation, as it does in the checks that come below?

Methods should not have to validate that they haven't been called with types that violate their declarations. Other validation (like whether it's `nil`), sure.

The code below this is also wrong; values with unknown types should be type checked before assigning to a variable declared as a specific type, not after. Either the values can only be `NSString`s, in which case the checks are pointless, or (more likely) they could be `NSString`s or `NSNull`s, in which case the variables should be declared `id`, not `NSString*`.